### PR TITLE
[BLO-1840] Determinism semantic search (first call)

### PR DESCRIPTION
**Pull request by rmuller-ml**
- Change rake-rs to allow "_", "-"and "/"
- Makes the keyword order deterministic
- Also change hyde temperature to 0

<a href="https://gitpod.io/#https://github.com/BloopAI/bloop/pull/1121"></a>

### CODE REVIEW

File: server/bleep/src/query/stopwords.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    })
+}
+
+fn phrases<'a>(phrases_iter: impl IntoIterator<Item = &'a str>) -> Vec<Vec<&'a str>> {
+    let phrases_iter = phrases_iter.into_iter();
+    let mut phrases = Vec::with_capacity(2 * phrases_iter.size_hint().0);
+    for s in phrases_iter.filter(|s| !s.is_empty()) {
+        let mut phrase = Vec::new();
+        for word in s.split_whitespace() {
+            if stop_words().contains(&word.to_lowercase()) {
```
Comment (rmuller-ml) [inline server/bleep/src/query/stopwords.rs:56]: maybe we should initialize stop_words once outside the for loops using a single call to stop_words()

File: server/bleep/src/query/stopwords.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+fn stop_words() -> &'static StopWords {
+    STOPWORDS.get_or_init(|| {
+        let mut sw = StopWords::new();
+        for w in STOP_WORDS_LIST.lines() {
+            sw.insert(w.to_string());
+        }
+        sw
+    })
+}
```
Comment (nerdypepper) [inline server/bleep/src/query/stopwords.rs:48]:
```suggestion
static STOPWORDS: Lazy<StopWords> = Lazy::new(|| {
    let word_list = include_str!("stopwords.txt");
    let mut sw = StopWords::new();
    for w in word_list.lines() {
        sw.insert(w);
    }
    sw
});
```
better expressed with `once_cell::sync::Lazy`.

File: server/bleep/src/query/stopwords.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+*/
+
+use lazy_regex::regex;
+use once_cell::sync::OnceCell;
+use std::collections::HashSet;
+
+type StopWords = HashSet<String>;
```
Comment (nerdypepper) [inline server/bleep/src/query/stopwords.rs:36]:
```suggestion
type StopWords = HashSet<&'static str>;
```
do we need these to be allocated strings?

File: server/bleep/src/query/stopwords.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    STOPWORDS.get_or_init(|| {
+        let mut sw = StopWords::new();
+        for w in STOP_WORDS_LIST.lines() {
+            sw.insert(w.to_string());
+        }
+        sw
+    })
+}
+
+fn phrases<'a>(phrases_iter: impl IntoIterator<Item = &'a str>) -> Vec<Vec<&'a str>> {
```
Comment (nerdypepper) [inline server/bleep/src/query/stopwords.rs:50]: can we add a doc-comment explaining what this does?

**Pull request merged**